qa/tasks/deepsea.py

```diff
@@ -124,6 +125,46 @@ def __init__(self, ctx, config):
         # self.log.debug("ctx.config {}".format(ctx.config))
         # self.log.debug("deepsea context: {}".format(deepsea_ctx))
 
+    def _deepsea_minions(self):
+        """
+        Set deepsea_minions pillar value
+        """
+        echo_cmd = (
+            'echo "deepsea_minions: \'*\'" > '
+            '/srv/pillar/ceph/deepsea_minions.sls'
+        )
+        self.master_remote.run(args=[
```

Comment:
Here and later you can use `self.master_remote.sh` instead.

Reply:
You can basically do the same with command: 

```
self.master_remote.sh(
    'cat | sudo tee /srv/pillar/ceph/deepsea_minions.sls', 
    stdin="deepsea_minions: \'*\'")
```

Reply:
There is a crucial difference between remote.run and remote.sh - the latter buffers output while the former does not. That means, if a command produces partial output and then times out, `remote.run` will display the partial output but `remote.sh` will not. 

Therefore, I only use remote.sh when I need to grep the output.

Reply:
@smithfarm I didn't get which partial output you mean, could you please give an example? The remote.sh is call remote.run and supposed to act the same way as remote.run.

Reply:
@smithfarm Oh, so, what about:
```
self.master_remote.run(args=
    'sudo tee /srv/pillar/ceph/deepsea_minions.sls', 
    stdin="deepsea_minions: \'*\'")
```

Reply:
@kshtsk remote.run also eats partial output if you use it to capture the output. By partial output, I mean if command "foobar" returns full output "foo bar" with possibility of a timeout between "foo" and "bar", then `remote.run(args="foobar")` will display "foo" in the log, whereas `remote.sh("foobar")` will not display anything in the log.

Reply:
added optimization to wip-qa-cleanup for testing

Reply:
> @kshtsk remote.run also eats partial output if you use it to capture the output. By partial output, I mean if command "foobar" returns full output "foo bar" with possibility of a timeout between "foo" and "bar", then `remote.run(args="foobar")` will display "foo" in the log, whereas `remote.sh("foobar")` will not display anything in the log.

so, where in the following example, there a 'foobar' is not displayed in the log:
> ```
> >>> from teuthology.orchestra.remote import Remote as R
> >>> r=R('127.0.0.1')
> >>> r.sh('cat', input="Hello world")
> 2018-11-13 19:18:28,471.471 INFO:teuthology.orchestra.run.127.0.0.1:Running:
> 2018-11-13 19:18:28,472.472 INFO:teuthology.orchestra.run.127.0.0.1:> cat
> 2018-11-13 19:18:28,523.523 INFO:teuthology.orchestra.run.127.0.0.1.stdout:Hello world
> 'Hello world'
> ```

I really don't understand your point, the `remote.sh(cmd, ...)` is just shortcut for `remote.run(args=cmd, stdout=cStreamIO(), ...)` and behavior is absolutely equivalent but the  return value.

Reply:
Your example is too simplistic. The code `r.sh('cat', input="Hello world")` does not generate a timeout between "Hello" and "world".

Your statement that the behavior is absolutely equivalent only holds for very simplistic commands that complete in a very short time.

It's easy to see the difference in the behavior, though, using a slightly more complicated scenario:

First, make the following script on the remote and call it `test.sh`:

```
#!/bin/sh -ex
for i in {1..10000} ; do sleep $i ; done
```

Then, run it via `timeout 60m test.sh`, first using `remote.sh('test.sh')` and then using `remote.run(args='test.sh')`. To see the difference, you have to do this using teuthology and watch the teuthology.log in real time (`tail -f /var/log/teuthology.??.log`).

Reply:
The point here is that we need to run the remote commands in a way that causes their output to be logged (to teuthology.log) immediately, not after the command completes - because we can't be sure it will complete at all.

Reply:
@smithfarm so when I ask you provide arguments to wait an HOUR and maybe something happens or not? Anyway your example is not convincing, it does not prove anything, because the same code can behave differently. The point you're trying to refer is probably relating to flushing buffer. Observing teuthology.log is not related to the remote.sh or remote.run.  However the original reference to 'it eats logs' is not true. If so, it is a bug which MUST be fixed, correspondingly IF it is really true you're trying to workaround it.

Reply:
> The point here is that we need to run the remote commands in a way that causes their output to be logged (to teuthology.log) immediately, not after the command completes - because we can't be sure it will complete at all.

You can't guarantee this, even with `remote.run` if the script you running does not flush buffers correctly. Again, `remote.sh` is equivalent to `remote.run` in sense of logging, because it calls `remote.run`.

Reply:
@smithfarm I can agree that it is reasonable to use run when you need to do some specific task, and you don't usually need to capture log, also, I would not use `remote.sh` when it comes to use long running scripts which generate huge amount of data. But in case, like this one, cat a file or copy some stuff, when it should return immediately, it is absolutely reasonable to use `remote.sh`.